src/lib/storage/dictionary_segment.hpp

```diff
@@ -0,0 +1,159 @@
+#pragma once
+
+#include <limits>
+#include <memory>
+#include <iostream>
+#include <string>
+#include <utility>
+#include <vector>
+#include <iterator>
+#include <set>
+
+#include "all_type_variant.hpp"
+#include "type_cast.hpp"
+#include "types.hpp"
+#include "base_attribute_vector.hpp"
+#include "base_segment.hpp"
+#include "fixed_size_attribute_vector.hpp"
+
+namespace opossum {
+
+// Even though ValueIDs do not have to use the full width of ValueID (uint32_t), this will also work for smaller ValueID
+// types (uint8_t, uint16_t) since after a down-cast INVALID_VALUE_ID will look like their numeric_limit::max()
+constexpr ValueID INVALID_VALUE_ID{std::numeric_limits<ValueID::base_type>::max()};
+
+// Dictionary is a specific segment type that stores all its values in a vector
+template <typename T>
+class DictionarySegment : public BaseSegment {
+ public:
+  /**
+   * Creates a Dictionary segment from a given value segment.
+   */
+  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment) :
+  _dictionary(std::make_shared<std::vector<T>>()) {
+    size_t attribute_vector_size = 0;
+    std::set<T> dictionary_helper;
+    for (size_t segment_iterator = 0; segment_iterator < base_segment->size(); ++segment_iterator) {
+      dictionary_helper.insert(type_cast<T>((*base_segment)[segment_iterator]));
+      attribute_vector_size++;
+    }
+    _dictionary->reserve(dictionary_helper.size());
+    for (auto it = dictionary_helper.begin(); it != dictionary_helper.end(); ) {
+      _dictionary->emplace_back(std::move(dictionary_helper.extract(it++).value()));
```

Comment:
Iterating in the loop header does not work here since the extract+move pair breaks the previous iterator.  Therefore we need to use postincrement here, if we want to get "maximum" performance (At least maximum afaik)

Reply:
Maybe use a range based loop? Would that work here?

Reply:
@Bensk1 I have a question for you here and will probably come by later. Afaik the constructor that uses iterators does not move the memory but rather creates copies, which would tank the performance here (especially for strings, etc. ), right? Is there a constructor that moves the memory?

Reply:
You are right, forget what I said. The constructor would copy the elements, you are moving them.